### PR TITLE
[RHCLOUD-20340] Config/Secret restarter

### PR DESCRIPTION
**Pull request by psav**

### CODE REVIEW

File: controllers/cloud.redhat.com/hashcache/hashcache.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+			delete(v.ClowdApps, types.NamespacedName{
+				Name:      obj.GetName(),
+				Namespace: obj.GetNamespace(),
+			})
+		}
+	}
+}
+
+func (hc *HashCache) CreateOrUpdateObject(obj client.Object) (bool, error) {
+	// NEED TO PASS IN Q HERE TO GET IT TO ALSO TRIGGER
```
Comment (adamrdrew) [inline controllers/cloud.redhat.com/hashcache/hashcache.go:104]: Is this comment still operative? I don't know what it means but it is in all caps which makes it look very important. If I encountered this while working I wouldn't know what it meant.

File: controllers/cloud.redhat.com/handlers.go [hunk 2]
```diff
@@ -111,7 +155,58 @@ func (e *enqueueRequestForObjectCustom) Create(evt event.CreateEvent, q workqueu
...
+	if evt.ObjectNew.GetAnnotations()["qontract.reconcile"] == "true" {
+		shouldUpdate, err := e.updateHashCacheForConfigMapAndSecret(evt.ObjectNew)
+		e.logMessage(evt.ObjectNew, "debug", fmt.Sprintf("shouldUpdate %s %v", e.ctrlName, shouldUpdate), getNamespacedName(evt.ObjectNew))
+		if err != nil {
+			e.logMessage(evt.ObjectNew, err.Error(), "", getNamespacedName(evt.ObjectNew))
+		}
+		if shouldUpdate {
+			_ = e.doUpdateToHash(evt.ObjectNew, q)
+		}
+		if evt.ObjectOld.GetAnnotations()["qontract.reconcile"] != evt.ObjectNew.GetAnnotations()["qontract.reconcile"] {
```
Comment (adamrdrew) [inline controllers/cloud.redhat.com/handlers.go:200]: Same comment as above - I'm surprised to see a specific annotation baked into the implementation. Why is this carved out in so many places? What makes this annotation special?

**Review by adamrdrew — DISMISSED** [review]
This is a big change set that adds complex new functionality, and is deeper and more sophisticated than the work I've done,  so it is challenging for me to review - but I think it looks good. You've got good test coverage both unit and e2e, and it looks like you've got all functionality and the API covered. I asked a few questions, but ultimately I trust your judgement.

File: controllers/cloud.redhat.com/handlers.go [hunk 3]
```diff
@@ -102,7 +119,34 @@ func (e *enqueueRequestForObjectCustom) logMessage(obj client.Object, msg string
 	logMessage(e.logr, "Reconciliation trigger", "ctrl", e.ctrlName, "type", msg, "resType", gvk.Kind, "sourceObj", fmt.Sprintf("%s/%s/%s", gvk.Kind, obj.GetNamespace(), obj.GetName()), "destObj", fmt.Sprintf("%s/%s/%s", toKind, own.Namespace, own.Name))
 }
 
+func (e *enqueueRequestForObjectCustom) updateHashCacheForConfigMapAndSecret(obj client.Object) (bool, error) {
+	switch obj.(type) {
+	case *core.ConfigMap, *core.Secret:
+		if obj.GetAnnotations()["qontract.reconcile"] == "true" {
```
[the remark this answers is not in the document]
Comment (psav) [reply controllers/cloud.redhat.com/handlers.go:125]: It's documented and is the app-sre "current" way of getting the functioanlity we broke by giving people clowder - this means a net no change for people

File: controllers/cloud.redhat.com/handlers.go (hunk 2 above)
Comment (psav) [reply controllers/cloud.redhat.com/handlers.go:200]: see above

**Review by maknop — APPROVED** [review]

**Pull request merged**